### PR TITLE
fix(aws-s3): Add validation for transitions in lifecycleRules

### PR DESCRIPTION
**Pull request by Damon1024**
Closes: #22103

Added validation for S3 bucket transition lifecycleRules

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by Naumel** [comment]
Hello,

Thank you for your time and submitting changes, please review https://www.conventionalcommits.org/en/v1.0.0/#summary for conventions that apply to [pull request titles](https://github.com/aws/aws-cdk/blob/v1-main/CONTRIBUTING.md#step-4-pull-request).

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 1]
```diff
@@ -1873,6 +1873,16 @@ export class Bucket extends BucketBase {
       throw new Error("Cannot use 'noncurrent' rules on a nonversioned bucket");
     }
 
+    rule.transitions?.forEach((transition) => {
+      if (transition.transitionAfter === undefined && transition.transitionDate === undefined) {
+        throw new Error('Exactly one of transitionDate and transitionAfter must be specified');
```
Comment (Naumel) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1878]: If there are two checks, please add customized messaging on what has gone wrong.

File: packages/@aws-cdk/aws-s3/test/rules.test.ts [hunk 2]
```diff
@@ -96,6 +96,38 @@ describe('rules', () => {
     });
   });
 
+  test('Bucket with no transition duration fails', () => {
```
Comment (Naumel) [inline packages/@aws-cdk/aws-s3/test/rules.test.ts:99]: The rule of thumb is to also have integ tests accompany this kind of changes.

**Review by Naumel — CHANGES_REQUESTED** [review]

**Comment by Naumel** [comment]
@Mergifyio update

File: packages/@aws-cdk/aws-s3/lib/bucket.ts (hunk 1 above)
Comment (Damon1024) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: I combined the two checks into one to match the real error message.

File: packages/@aws-cdk/aws-s3/test/rules.test.ts (hunk 2 above)
Comment (Damon1024) [reply packages/@aws-cdk/aws-s3/test/rules.test.ts:99]: sorry, AFAIK I cannot do an integ test for this.

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 3]
```diff
@@ -1874,11 +1874,10 @@ export class Bucket extends BucketBase {
...
 
     rule.transitions?.forEach((transition) => {
-      if (transition.transitionAfter === undefined && transition.transitionDate === undefined) {
-        throw new Error('Exactly one of transitionDate and transitionAfter must be specified');
-      }
-
-      if (transition.transitionAfter !== undefined && transition.transitionDate !== undefined) {
+      if (
+        (transition.transitionAfter === undefined && transition.transitionDate === undefined) ||
+        (transition.transitionAfter !== undefined && transition.transitionDate !== undefined)
```
Comment (Naumel) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1879]: The condition can be simplified and made to check for empty entries too:
```
((transition.transitionAfter && transition.transitionDate) || (!transition.transitionAfter && !transition.transitionAfter))
```
***But***, being verbose makes it easier to troubleshoot; if the message states clearly you either have none or both specified.

**Review by Naumel — DISMISSED** [review]

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 4]
```diff
@@ -1873,6 +1873,12 @@ export class Bucket extends BucketBase {
       throw new Error("Cannot use 'noncurrent' rules on a nonversioned bucket");
     }
 
+    rule.transitions?.forEach((transition) => {
+      if ((transition.transitionAfter && transition.transitionDate) || (!transition.transitionAfter && !transition.transitionDate)) {
+        throw new Error('Exactly one of transitionDate and transitionAfter must be specified');
+      }
+    });
+
```
Comment (Naumel) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:1881]: I take it you really dislike throwing separate Errors with dedicated messages for the cases when either both are specified or both are not - is there a particular reason for that?

Comment (Damon1024) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: I split it into two cases with different messages.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
I actually think the need for this sort of validation reveals that this is the wrong user experience in general.  The correct fix here is not throwing errors, but to deprecate these properties and implement something like what we have in [Schedule](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-events/lib/schedule.ts).

**Pull request closed without merging**